contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,268 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.10;
+import "../libraries/Address.sol";
+
+///  ========== TYPES ==========
+import "../types/BaseAllocator.sol";
+
+///  ========== INTERFACES ==========
+import "./interfaces/IMasterChef.sol";
+import "./interfaces/ISushiBar.sol";
+
+struct PoolData {
+    bool poolActive;
+    uint256 id;
+}
+
+/**
+ *  @notice Contract deploys liquidity from treasury into the Onsen program,
+ *  earning $SUSHI that can be staked and/or deposited into the treasury.
+ */
+contract OnsenAllocatorV2 is BaseAllocator {
+    using SafeERC20 for IERC20;
+
+    /// ========== STATE VARIABLES ==========
+
+    address public sushi; /// $SUSHI token
+    address public xSushi; /// $xSUSHI token
+    address public masterChef; /// Onsen contract
+
+    address public treasury; /// Olympus Treasury
+
+    mapping(address => PoolData) internal _lpToOnsenId;
+
+    /// ========== CONSTRUCTOR ==========
+
+    constructor(
+        address _chef,
+        address _sushi,
+        address _xSushi,
+        address _treasury,
+        AllocatorInitData memory data
+    ) BaseAllocator(data) {
+        require(_chef != address(0));
+        masterChef = _chef;
+        require(_sushi != address(0));
+        sushi = _sushi;
+        require(_xSushi != address(0));
+        xSushi = _xSushi;
+        require(_treasury != address(0));
+        treasury = _treasury;
+
+        /**  @dev approve for safety, sushi is being staked and xsushi is being instantly sent to treasury and that is fine
+         *   but to be absolutely safe this one approve won't hurt
+         */
+        IERC20(sushi).approve(address(extender), type(uint256).max);
+        IERC20(xSushi).approve(address(extender), type(uint256).max);
+    }
+
+    /**
+     * @notice Find out which LP tokens to be deposited into onsen, then deposit it into onsen pool and stake the sushi tokens that are returned.
+     * 1- Based on the LP token to be deposited, find out the Onsen Pool Id
+     * 2- Deposit LP token into onsen, if deposit is succesfull we should get in return sushi tokens
+     * 2- Stake the Sushi rewards tokens returned from deposit LP tokens and as rewards form onsen LP stake, we should get in return xSushi tokens
+     * 3- Keep the xSushi on the allocator contract for easily deallocation
+     * 4- Calculate gains/loss based on the LP balance on the Onsen Pool.
+     */
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        uint256 index = tokenIds[id];
+        IERC20 LPtoken = _tokens[index];
+        uint256 balance = LPtoken.balanceOf(address(this));
+
+        /// Deposit LP token into onsen, if deposit succesfull this address should have in return sushi tokens
+        if (balance > 0) {
```

Comment:
so i read the masterchef contract and it claims your pending sushi when u deposit or withdraw correct?

Not sure what the strategy is financially but i would imagine we have a case where we would wana not deposit or withdraw anything but compound our stack by claiming and staking sushi. 

If we want to do that potentially maybe we can remove the balance > 0 check.

Reply:
+1

Reply:
"Update" function updated so if there is no LP token balance, then harvest the sushi token rewards so those can be staked.

@Paul4912 @PARABRAHMAN0 , please check the current solution and let me know if you think this is fine.

Reply:
yea that works too